### PR TITLE
#56 HackerRank | Algorithms > Sorting > Insertion Sort - Part 1

### PR DESCRIPTION
**Pull request by fedotkin**
Tested [Insertion Sort - Part 1](https://www.hackerrank.com/challenges/insertionsort1/submissions/code/314858790)

### CODE REVIEW

File: hackerrank/test/Algorithms/Sorting/InsertionSortPart1_ResultTests.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    [Fact]
+    public void InsertionSort1_Example_PrintedArrays()
+    {
+        // Arrange
+        int arrayLength = 5;
+        List<int> array = new List<int>() { 1, 2, 4, 5, 3 };
+
+        var asm = Assembly.GetExecutingAssembly();
+        var dir = Path.GetDirectoryName(asm.Location);
+        string filePath = Path.Combine(dir, nameof(InsertionSort1_Example_PrintedArrays) + ".txt");
```
Comment (raman-m) [inline hackerrank/test/Algorithms/Sorting/InsertionSortPart1_ResultTests.cs:18]: For unit tests it does not matter the real folder of reading/writing streams.
So, you can use `Environment.CurrentDirectory` property without problems.

File: hackerrank/test/Algorithms/Sorting/InsertionSortPart1_ResultTests.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    [Fact]
+    public void InsertionSort1_Sample_PrintedArrays()
+    {
+        // Arrange
+        int arrayLength = 5;
+        List<int> array = new List<int>() { 2, 4, 6, 8, 3 };
+
+        var asm = Assembly.GetExecutingAssembly();
+        var dir = Path.GetDirectoryName(asm.Location);
+        string filePath = Path.Combine(dir, nameof(InsertionSort1_Sample_PrintedArrays) + ".txt");
```
Comment (raman-m) [inline hackerrank/test/Algorithms/Sorting/InsertionSortPart1_ResultTests.cs:44]:
```diff
- Don't use assembly-related operations inside of unit test! 
```

**Review by raman-m — CHANGES_REQUESTED** [review]
Fix issues please!

File: hackerrank/src/Algorithms/Sorting/InsertionSortPart1.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        int index = n - 1;
+        int movingElement = arr[index];
+
+        while (index > 0 && (arr[index - 1] > movingElement))
+        {
+            arr[index] = arr[index - 1];
+            Console.WriteLine(string.Join(" ", arr));
+            index--;
+        }
```
[the remark this answers is not in the document]
Comment (fedotkin) [reply hackerrank/src/Algorithms/Sorting/InsertionSortPart1.cs:15]: Well, also `index` and `movingElement` variables work only in the loop, but they are also necessary after the loop for the correct termination of the program. 
For example, input: `10, [ 2, 3, 4, 5, 6, 7, 8, 9, 10, 1 ]`

[the remark this answers is not in the document]
Comment (raman-m) [reply hackerrank/src/Algorithms/Sorting/InsertionSortPart1.cs:15]: The `i` and `movingElement` variables will be visible after **for**-loop also.

[the remark this answers is not in the document]
Comment (fedotkin) [reply hackerrank/src/Algorithms/Sorting/InsertionSortPart1.cs:15]: `i` variable is also needed outside of the loop

[the remark this answers is not in the document]
Comment (fedotkin) [reply hackerrank/src/Algorithms/Sorting/InsertionSortPart1.cs:15]: sorry, the code doesn't work

[the remark this answers is not in the document]
Comment (raman-m) [reply hackerrank/src/Algorithms/Sorting/InsertionSortPart1.cs:15]: You need to explain shortly why it doesn't work!
To extand the scope of the `index` and `movingElement` variables out from the **for**-loop scope you can apply the short declaration life hack. It allows to make `index` and `movingElement` visible in whole scope of the method.
```csharp
int index = 0, movingElement = 0; // life hack!
for (index = n - 1, movingElement = arr[index]; // initialization
    index > 0 && arr[index - 1] > movingElement; // condition
    index--) // iteration
//for (index = n - 1, movingElement = arr[index]; index > 0 && arr[index-1] > movingElement; index--)
{
    arr[index] = arr[index - 1];
    Console.WriteLine(string.Join(" ", arr));
}
arr[index] = movingElement;
Console.WriteLine(string.Join(" ", arr));
```
This is your algorithm based on **for**-loop.
Enjoy!

**Review by raman-m — APPROVED** [review]
```diff
+ Approved
```

**Pull request merged**